### DISCUSSION TITLE
Unique indexes on link endpoints

### DISCUSSION THREAD

**Discussion by CodesInChaos** [geldata/gel#1714] [Q&A] [answered]
I understand that referencing arbitrary properties of a link source/target is problematic, since they might be in different tables. But the ids of the source/target itself should be available on the link (as foreign key or because the link is embedded into the table). Being able to reference them in a unique index is needed for common scenarios:

1. Array-like multi-links need an unique constraint on `(source.id, link.index)`
2. Allowing only one link between a pair of nodes needs one on `(source.id, target.id)`

**Comment by elprans** [comment, marked as answer]
> Array-like multi-links need an unique constraint on `(source.id, link.index)`

This is already the case by default.

> Allowing only one link between a pair of nodes needs one on `(source.id, target.id)`

Use `constraint exclusive` for this.

**Reply by CodesInChaos** [reply]
> 
> 
> > Array-like multi-links need an unique constraint on `(source.id, link.index)`
> 
> This is already the case by default.

There is build-in support for ordered links (which behaves like an array)? I only know of links which behave like (multi)-sets and don't preserve an order.

> 
> > Allowing only one link between a pair of nodes needs one on `(source.id, target.id)`
> 
> Use `constraint exclusive` for this.

I thought that's equivalent to an exclusive constraint on the target, not on the source-target pair? I'm thinking about a graph where each node can have multiple links, but there can only be one link between a pair of nodes. e.g. a student can be in multiple classes, a class can contain multiple students, but a student can't be twice in the same class.

**Reply by elprans** [reply]
> but there can only be one link between a pair of nodes

Oh, that's enforced by default as well.

**Reply by elprans** [reply]
> There is build-in support for ordered links

There isn't, but it's fairly easy to implement via a link property:

```
type Entry {
    multi link tags -> Tag {
        property order -> int64;
    }
}
```

```
INSERT Entry {
    tags := (
       FOR tag IN {enumerate($tags)}
       SELECT Tag { @order := tag.0 } FILTER .name = tag.1
    )
}
```

**Reply by CodesInChaos** [reply]
> > but there can only be one link between a pair of nodes
> 
> Oh, that's enforced by default as well.

So links are not multi-sets by default? I assume if I want to allow that I have to define a type serving as linker table like in a relational database?

**Reply by CodesInChaos** [reply]
> > There is build-in support for ordered links
> 
> There isn't, but it's fairly easy to implement via a link property:
> 
> ```
> type Entry {
>     multi link tags -> Tag {
>         property order -> int64;
>     }
> }
> ```

That's what I use at the moment, but I'm trying to enforce the invariant that there are no two links originating at the same `Entry` which have the same `order`. Which would require a unique index on the `(Entry.id, order)` tuple.

**Reply by elprans** [reply]
This should work:

```
type Entry {
    multi link tags -> Tag {
        property order -> int64;
        constraint exclusive on 
            ((__subject__@source, __subject__@order));
    }
}
```

**Reply by elprans** [reply]
> So links are not multi-sets by default?

They aren't.

> I assume if I want to allow that I have to define a type serving as linker table like in a relational database?

That's correct.

**Reply by CodesInChaos** [reply]
`__subject__@source` that's what I was looking for. Is there a similar predefined property for the target?

**Reply by elprans** [reply]
> Is there a similar predefined property for the target?

There is: `@target`.  It seems it's somewhat broken in link constraints though
